internal/controllers/ipaddressclaim.go

```diff
@@ -163,26 +164,77 @@ func (r *IPAddressClaimReconciler) reconcile(ctx context.Context, c *ipamv1.IPAd
 
 	log = log.WithValues("pool name", pool.GetName())
 
-	address := poolutil.AddressByName(addresses, c.Name)
+	address := poolutil.AddressByName(addresses, claim.Name)
 	if address == nil {
 		var err error
-		address, err = r.allocateAddress(ctx, c, pool, addresses)
+		address, err = r.allocateAddress(ctx, claim, pool, addresses)
 		if err != nil {
 			log.Error(err, "failed to allocate address")
 			return ctrl.Result{}, err
 		}
+
+		log.Info("Allocated IPAddress for IPAddressClaim",
+			"IPAddressClaim", fmt.Sprintf("%s/%s", claim.Namespace, claim.Name),
+			"IPAddress", fmt.Sprintf("%s/%s %s", address.Namespace, address.Name, address.Spec.Address),
+			"InClusterIPPool", pool.GetName())
+	} else {
+		if err := r.ensureOwnerRefsAndFinalizer(ctx, address, claim, pool); err != nil {
+			log.Error(err, "failed to ensure owner refs and finalizer on address")
+			return ctrl.Result{}, err
+		}
+
+		log.Info("Found existing IPAddress for IPAddressClaim",
+			"IPAddressClaim", fmt.Sprintf("%s/%s", claim.Namespace, claim.Name),
+			"IPAddress", fmt.Sprintf("%s/%s %s", address.Namespace, address.Name, address.Spec.Address),
+			"InClusterIPPool", pool.GetName())
 	}
+
 	if !address.DeletionTimestamp.IsZero() {
 		// We prevent deleting IPAddresses while their corresponding IPClaim still exists since we cannot guarantee that the IP
 		// wil remain the same when we recreate it.
 		log.Info("Address is marked for deletion, but deletion is prevented until the claim is deleted as well.", "address", address.Name)
 	}
 
-	c.Status.AddressRef = corev1.LocalObjectReference{Name: address.Name}
+	claim.Status.AddressRef = corev1.LocalObjectReference{Name: address.Name}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *IPAddressClaimReconciler) ensureOwnerRefsAndFinalizer(ctx context.Context, address *ipamv1.IPAddress, claim *ipamv1.IPAddressClaim, pool genericInClusterPool) error {
```

Comment:
Should the logic be modified to check and update only if necessary? Also, maybe using a `PatchHelper` from the [CAPI utils](https://github.com/kubernetes-sigs/cluster-api/blob/323e257147a9a9094dde1215aada4e33a0788278/util/patch/patch.go#L52) be useful here.
You can find instances of this usage in the CAPI/CAPV codebases

Reply:
Updated.

We didn't realize the patch helper would avoid the API call if no mods were found. Today we learned.